src/4D_api.cpp

```diff
@@ -966,6 +966,423 @@ std::string pj_add_type_crs_if_needed(const std::string& str)
+ * constructed with (left, bottom, 180, top) and the second with
+ * (-180, bottom, top, right).
+ *
+ * When projecting from polar projections to geographic,
+ * lon, lat output order is required.
+ *
+ * @param P The PJ object representing the transformation.
+ * @param direction The direction of the transformation.
+ * @param left The input left bounding coordinate.
+ * @param bottom The input bottom bounding coordinate.
+ * @param right To input right bounding coordinate.
+ * @param top The input top bounding coordinate.
+ * @param out_left The output value for the left bounding coordinate.
+ * @param out_bottom The output value for the bottim bounding coordinate.
+ * @param out_right The output value for the right bounding coordinate.
+ * @param out_top The output value for the top bounding coordinate.
+ * @param densify_pts Recommended to use 21. This is the number of points
+ *     to use to densify the bounding polygon in the transformation.
+ * @return an integer. 1 if successful. 0 if failures encountered.
+ * @since 8.2
+ */
+int proj_trans_bounds(PJ *P,
+                      PJ_DIRECTION direction,
+                      const double left,
+                      const double bottom,
+                      const double right,
+                      const double top,
+                      double* out_left,
+                      double* out_bottom,
+                      double* out_right,
+                      double* out_top,
+                      int densify_pts
+) {
+    *out_left = HUGE_VAL;
+    *out_bottom = HUGE_VAL;
+    *out_right = HUGE_VAL;
+    *out_top = HUGE_VAL;
+
+    if (P == nullptr) {
+        proj_log_error(P, _("NULL P object not allowed,"));
+        proj_errno_set (P, PROJ_ERR_INVALID_OP_ILLEGAL_ARG_VALUE);
+        return false;
+    }
+    if (densify_pts < 0) {
+        proj_log_error(P, _("densify_pts must be greater than 0."));
+        proj_errno_set (P, PROJ_ERR_INVALID_OP_ILLEGAL_ARG_VALUE);
+        return false;
+    }
+
+    PJ_PROJ_INFO pj_info = proj_pj_info(P);
+    if (pj_info.id == nullptr) {
+        proj_log_error(P, _("NULL transformation not allowed,"));
+        proj_errno_set (P, PROJ_ERR_INVALID_OP_ILLEGAL_ARG_VALUE);
+        return false;
+    }
+    if (strcmp(pj_info.id, "noop") == 0) {
+        *out_left = left;
+        *out_right = right;
+        *out_bottom = bottom;
+        *out_top = top;
+        return true;
+    }
+
+    bool degree_output = proj_degree_output(P, direction) != 0;
+    bool degree_input = proj_degree_input(P, direction) != 0;
+    if (degree_output && densify_pts < 2) {
+        proj_log_error(P, _("densify_pts must be at least 2 if the output is geograpic."));
+        proj_errno_set (P, PROJ_ERR_INVALID_OP_ILLEGAL_ARG_VALUE);
+        return false;
+    }
+
+    int side_pts = densify_pts + 1;  // add one because we are densifying
+    int boundary_len = side_pts * 4;
+    std::vector<double> x_boundary_array(boundary_len);
```

Comment:
it could be prudent to do in 2 steps
```
std::vector<double> x_boundary_array;
std::vector<double> y_boundary_array;
try
{
    x_boundary_array.resize(boundary_len);
    t_boundary_array.resize(boundary_len);
}
catch( const std::exception & e ) // memory allocation failure
{
     // return an error using e.what()
}
```

Reply:
Thanks for the suggestion :+1: